src/main.rs

```diff
@@ -966,4 +994,102 @@ mod tests {
+            let os_disk = format!("file={},format=qcow2", guest.disks[0].as_str());
+            let cloud_init_disk = format!("file={},format=raw", guest.disks[1].as_str());
+
+            let vfio_tap0 = "vfio-tap0";
+            let vfio_tap1 = "vfio-tap1";
+
+            let ssh_net = "ssh-net";
+            let vfio_net = "vfio-net";
+
+            let netdev_ssh = format!(
+                "tap,ifname={},id={},script=no,downscript=no",
+                vfio_tap0, ssh_net
+            );
+            let netdev_ssh_device = format!(
+                "virtio-net-pci,netdev={},disable-legacy=on,iommu_platform=on,ats=on,mac={}",
+                ssh_net, guest.guest_mac
+            );
+
+            let netdev_vfio = format!(
+                "tap,ifname={},id={},script=no,downscript=no",
+                vfio_tap1, vfio_net
+            );
+            let netdev_vfio_device = format!(
+                "virtio-net-pci,netdev={},disable-legacy=on,iommu_platform=on,ats=on,mac={}",
+                vfio_net, guest.l2_guest_mac
+            );
+
+            let mut qemu_child = Command::new("qemu-system-x86_64")
+                .args(&["-machine", "q35,accel=kvm,kernel_irqchip=split"])
+                .args(&["-bios", &ovmf_path])
+                .args(&["-smp", "sockets=1,cpus=4,cores=2"])
+                .args(&["-cpu", "host"])
+                .args(&["-m", "1024"])
+                .args(&["-vga", "none"])
+                .args(&["-nographic"])
+                .args(&["-drive", &os_disk])
+                .args(&["-drive", &cloud_init_disk])
+                .args(&["-device", "virtio-rng-pci"])
+                .args(&["-netdev", &netdev_ssh])
+                .args(&["-device", &netdev_ssh_device])
+                .args(&["-netdev", &netdev_vfio])
+                .args(&["-device", &netdev_vfio_device])
+                .args(&[
+                    "-device",
+                    "intel-iommu,intremap=on,caching-mode=on,device-iotlb=on",
+                ])
+                .args(&["-fsdev", &vfio_9p_path])
+                .args(&[
+                    "-device",
+                    "virtio-9p-pci,fsdev=shared,mount_tag=cloud_hypervisor",
+                ])
+                .spawn()
+                .unwrap();
+
+            thread::sleep(std::time::Duration::new(30, 0));
+
+            guest.ssh_command_l1("sudo systemctl start vfio");
+            thread::sleep(std::time::Duration::new(30, 0));
+
+            // We booted our cloud hypervisor L2 guest with a "VFIOTAG" tag
+            // added to its kernel command line.
+            // Let's ssh into it and verify that it's there. If it is it means
+            // we're in the right guest (The L2 one) because the QEMU L1 guest
+            // does not have this command line tag.
+            aver_eq!(
+                tb,
+                guest
+                    .ssh_command_l2("cat /proc/cmdline | grep -c 'VFIOTAG'")
```

Comment:
That's awesome, I just read the test, it's simple but very powerful. By ssh'ing, we validate the network device `virtio-net` has actually been passed through properly.